src/color.rs

```diff
@@ -562,109 +602,232 @@ fn parse_color_function<'i, 't, ComponentParser>(
+        return legacy_syntax_result;
+    }
+
+    // Determine whether this is the number syntax or the percentage syntax.
+    let number_result: Result<
+        (f32, f32, f32),
+        ParseError<<ComponentParser as ColorComponentParser>::Error>,
+    > = arguments.try_parse(|input| {
+        let red_number = component_parser.parse_number(input)?;
+        let green_number = component_parser.parse_number(input)?;
+        let blue_number = component_parser.parse_number(input)?;
+
+        Ok((red_number, green_number, blue_number))
+    });
+
+    let computed_red;
+    let computed_green;
+    let computed_blue;
+    if number_result.is_ok() {
+        let (red_number, green_number, blue_number) = number_result?;
+
+        computed_red = clamp_floor_256_f32(red_number);
+        computed_green = clamp_floor_256_f32(green_number);
+        computed_blue = clamp_floor_256_f32(blue_number);
     } else {
-        green = clamp_unit_f32(component_parser.parse_percentage(arguments)?);
-        if uses_commas {
-            arguments.expect_comma()?;
-        }
-        blue = clamp_unit_f32(component_parser.parse_percentage(arguments)?);
+        let red_percentage = component_parser.parse_percentage(arguments)?;
+        let green_percentage = component_parser.parse_percentage(arguments)?;
+        let blue_percentage = component_parser.parse_percentage(arguments)?;
+
+        computed_red = clamp_unit_f32(red_percentage);
+        computed_green = clamp_unit_f32(green_percentage);
+        computed_blue = clamp_unit_f32(blue_percentage);
     }
 
-    Ok((red, green, blue, uses_commas))
+    let computed_alpha = clamp_unit_f32(parse_alpha_component(component_parser, arguments)?);
+
+    Ok((computed_red, computed_green, computed_blue, computed_alpha))
 }
 
-/// Parses hsl and hbw syntax, which happens to be identical.
-///
-/// https://drafts.csswg.org/css-color/#the-hsl-notation
-/// https://drafts.csswg.org/css-color/#the-hbw-notation
 #[inline]
-fn parse_hsl_hwb<'i, 't, ComponentParser>(
+fn fix_hue_rounding_errors(value: f32) -> f32 {
+    // Subtract an integer before rounding, to avoid some rounding errors.
+    (value - 360. * (value / 360.).floor()) / 360.
+}
+
+/// https://w3c.github.io/csswg-drafts/css-color-4/#the-hsl-notation
+#[inline]
+fn parse_rgb_components_hsl<'i, 't, ComponentParser>(
     component_parser: &ComponentParser,
     arguments: &mut Parser<'i, 't>,
-    to_rgb: impl FnOnce(f32, f32, f32) -> (f32, f32, f32),
-    allow_comma: bool,
-) -> Result<(u8, u8, u8, bool), ParseError<'i, ComponentParser::Error>>
+) -> Result<(u8, u8, u8, u8), ParseError<'i, ComponentParser::Error>>
 where
     ComponentParser: ColorComponentParser<'i>,
 {
-    // Hue given as an angle
-    // https://drafts.csswg.org/css-values/#angles
-    let hue_degrees = component_parser.parse_angle_or_number(arguments)?.degrees();
+    // Try to parse legacy syntax first.
+    let legacy_syntax_result = arguments.try_parse(|input| {
```

Comment:
This refactoring seems somewhat gratuituous? What advantages does it give us? At a glance it seems it would make parsing a bit slower / rewind more.